adapter/serviceregistry/aggregate/controller.go

```diff
@@ -47,22 +48,36 @@ func (c *Controller) AddRegistry(registry Registry) {
 }
 
 // Services lists services from all platforms
-func (c *Controller) Services() []*model.Service {
+func (c *Controller) Services() ([]*model.Service, error) {
 	services := make([]*model.Service, 0)
+	var errs error
 	for _, r := range c.registries {
-		services = append(services, r.Services()...)
+		svcs, err := r.Services()
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		}
+		services = append(services, svcs...)
 	}
-	return services
+	return services, errs
 }
 
 // GetService retrieves a service by hostname if exists
-func (c *Controller) GetService(hostname string) (*model.Service, bool) {
+func (c *Controller) GetService(hostname string) (*model.Service, error) {
+	var errs error
 	for _, r := range c.registries {
-		if service, exists := r.GetService(hostname); exists {
-			return service, true
+		service, err := r.GetService(hostname)
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		}
+		if service != nil {
+			if errs != nil {
+				glog.Warningf("GetService() found match but encountered an error: %v", errs)
+			}
+			return service, nil
```

Comment:
Why return nil instead of the error?

Reply:
It is a design decision that is up for debate.  Of the N registries, at least one of them experienced an error - but one of the registries found the service that was requested.  Does one broken registry effectively break the aggregator?

Reply:
What you describe is not the logic I see in the code. What I see is the situation when service was found in a registry, but an error was reported by the same registry. For this, we should return it, it is up to the caller of this API to decide if the service is valid (I would say it's never valid).

For the N registries: I am for returning the service if found and nil error, only if the error from the registry where the service was found is nil, otherwise you have to return the error as well.  But the current code does not distinguish between the case where the error was from previous iterations, or from the current iteration. You need to re-arrange the ifs logic.

Please add the comment to explain the above decision, and log the errors that would otherwise be swallowed.